### DISCUSSION TITLE
Audi MHI2 Languages after EU convertion

### DISCUSSION THREAD

**Discussion by kmaci27** [Mr-MIBonk/M.I.B._More-Incredible-Bash#574] [General]
Hello all,

Car is a Audi Q7 from USA With MHi2 unit. I did convert the unit using M.I.B 3.7.1(latest version) everything went smooth. I flashed the European FW that is suggested but i dont have EU Languages. I have installed EU maps, but i dont have EU languages It only shows English and Spanish (mexican flag)Anyone can help me with that?  Thank you

**Comment by filippz** [comment]
I had similar problems with (unconverted) MHI2_EU_AUG22 unit where I used M.I.B. to try and fix some SDS issue only to make it worse as it seems _setlang_ script has wrong dataset for those units. It seems that MHI2Q_EU_AUG22 units also suffer form the same problem so I have prepared "fixed" _setlang_ with language datasets from actual (unconverted) MHI2_EU_AUG22 and MHI2Q_EU_AUG22 units. You can try by replacing M.I.B. _setlang_ in _apps_ folder of the SD card and running GEM and selecting _m.i.b._ -> _multimedia_system_ -> _languages_
Please be advised that this **_setlang_ HAS NOT BEEN TESTED** on MHI2Q units and I obtained MHI2Q dataset from a group member so those values might not be proper and/or script might be broken.
[setlang.zip](https://github.com/user-attachments/files/17509785/setlang.zip)

**Reply by kmaci27** [reply]
Thank you. The worse scenario that might happen lol?

**Reply by filippz** [reply]
I guess you could brick your unit and require Telnet/UART access to recover it.
It's up to you - but you should have a backup of the existing dataset (MU1329-datasets.csv) so reverting should be easy. You can also try running GEM and selecting _m.i.b._ -> _multimedia_system_ -> _languages_ with original _setlang_ and only if that doesn't help try the one I've attached earlier.

**Reply by Nikola034** [reply]
> I had similar problems with (unconverted) MHI2_EU_AUG22 unit where I used M.I.B. to try and fix some SDS issue only to make it worse as it seems _setlang_ script has wrong dataset for those units. It seems that MHI2Q_EU_AUG22 units also suffer form the same problem so I have prepared "fixed" _setlang_ with language datasets from actual (unconverted) MHI2_EU_AUG22 and MHI2Q_EU_AUG22 units. You can try by replacing M.I.B. _setlang_ in _apps_ folder of the SD card and running GEM and selecting _m.i.b._ -> _multimedia_system_ -> _languages_ Please be advised that this **_setlang_ HAS NOT BEEN TESTED** on MHI2Q units and I obtained MHI2Q dataset from a group member so those values might not be proper and/or script might be broken. [setlang.zip](https://github.com/user-attachments/files/17509785/setlang.zip)

Поздрав Филипе, 

занима ме да ли је могуће на MHI2_EU јединицу убацити српски језик на тај начин....убацим овај твој setlang и онда га кроз M.I.B селектујем.

Хвала унапред на одговору!

**Reply by filippz** [reply]
Pozdrav/Hello,

it's not possible just by changing this dataset because the actual translations are in `lang_data.zip` contained in the firmware itself and sadly both MHI2_EU_AUG22 and MHI2Q_EU_AUG22 units don't have _hr_HR_ or _sr_RS_ in them - the closest (ExYU close) match is _sl_SI_ in MHI2Q. Some other languages like _ro_RO_ exists only in MHI2Q so adding them to MHI2 could be fairly easy as a lot of strings (but not all) exists in MHI2Q that MHI2 needs.

That being said, some time ago I've written an application that can be used change/add new languages to `lang_data.zip` but it would either require someone to translate all the strings manually (I'm too lazy for that myself) or to find some MHI2 firmware that contains either _hr_HR_ or _sr_RS_ in them and use them as reference for adding translation to MHI2/MHI2Q. More details are here: https://github.com/Mr-MIBonk/M.I.B._More-Incredible-Bash/discussions/441.
Here you can see how would it look to try translating to a new _hr_HR_ language that has been copied from _en_GB_ by using MHI2Q _si_SL_ translations:
![image](https://github.com/user-attachments/assets/12bfa540-4dc0-40ae-a19f-8cad7976a5b8)


While I haven't actually added a new language as a proof of concept I've changed one of the strings and replaced `lang_data.zip` on my MHI2 device:
![image](https://github.com/user-attachments/assets/a7621531-f5c3-4b95-a4b7-85b37ad42e70)

**Comment by kmaci27** [comment]
Thank you
